Simplify arguments list of ull_adv_sync_pdu_alloc

ppryga-nordic opened the pull request:
The PR provides enhancement to use of ull_adv_sync_pdu_alloc.

The function had:
- Arguments that were related with updating of extended advertising PDU header. The arguments were actually used to manage allocation of `extra_data` memory.
- Unused argument `hdr_data`.

ppryga-nordic committed:
Bluetooth: controller: simplify arg list of ull_adv_sync_pdu_alloc

ull_adv_sync_pdu_alloc function was extracted from ull_adv_sync_pdu_-
-set_clear function. The arguments list of ull_adv_sync_pdu_alloc
was derived from source function but the acutal functionality is
only part of what it was before.

The ull_adv_sync_pdu_alloc does not change extended advertising
PDU header fileds. The hdr_add_fields and hdr_rem_fields arguments
were used to allocate or not allocate memory for extra_data related
with new periodic advertising PDU. Also hrd_data pointer was
not used by the function.

The function arguments list was simplified.
New enum ull_adv_pdu_extra_data_flag was introduced to provide
named flags for extra_data memory management.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: remove extra_data use in BIG create and terminate

extra_data is not used by ll_big_create and ll_big_terminate.
extra_data_prev and extra_data_new are optional arguments for
ull_adv_sync_pdu_alloc.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak reviewed (approved):
No objections from me.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -795,9 +789,8 @@ uint8_t ull_adv_sync_pdu_alloc(struct ll_adv_set *adv,
-	if ((hdr_add_fields & ULL_ADV_PDU_HDR_FIELD_CTE_INFO) ||
-	    (!(hdr_rem_fields & ULL_ADV_PDU_HDR_FIELD_CTE_INFO) &&
-	     ed_prev)) {
+	if (extra_data_flag == ULL_ADV_PDU_EXTRA_DATA_ALLOC_ALWAYS ||
+	    (!(extra_data_flag == ULL_ADV_PDU_EXTRA_DATA_ALLOC_NEVER) && ed_prev)) {
```
Semantically I think this makes more sense:
```suggestion
	if (extra_data_flag == ULL_ADV_PDU_EXTRA_DATA_ALLOC_ALWAYS ||
	    (extra_data_flag == ULL_ADV_PDU_EXTRA_DATA_ALLOC_IF_EXIST && ed_prev)) {
```

ppryga replied:
Changed according to suggestion. Thanks!

ppryga-nordic committed:
Bluetooth: controller: remove unnecessary ull_adv_ext_hdr_data struct

After extracting from ull_adv_sync_pdu_set_clear functions:
ull_adv_sync_pdu_alloc, ull_adv_sync_extra_data_set_clear
Use of ull_adv_ext_hrd_data structure became unnecessary.
The extra_data member of the structure was never used
due to separate function responsible for setting extra_data
content.
In all cases content for extended advertising header fields
may be passed to the ull_adv_sync_pdu_set_clear directly
as void * pointer, instead of pointer to ull_adv_ext_hdr_data.

This simplifies use of ull_adv_sync_pdu_set_clear as well
as removes unneccesary code to handle special structure type.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@cvinayak @Thalley I've added another commit to the PR. It removes `struct ull_adv_ext_hdr_data` because it is no longer needed.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -185,12 +174,9 @@ uint8_t ull_adv_sync_pdu_alloc(struct ll_adv_set *adv,
-				   uint16_t hdr_add_fields,
-				   uint16_t hdr_rem_fields,
-				   struct ull_adv_ext_hdr_data *hdr_data);
+uint8_t ull_adv_sync_pdu_set_clear(struct lll_adv_sync *lll_sync, struct pdu_adv *ter_pdu_prev,
+				   struct pdu_adv *ter_pdu, uint16_t hdr_add_fields,
+				   uint16_t hdr_rem_fields, void *hdr_data);
```
It actually seems to me that the last 3 arguments (`hdr_add_fields`, `hdr_rem_fields` and `hdr_data`) could benefit from being a struct, or at least `hdr_add_fields` and `hdr_data` as, if I understand correct, `hdr_data`, is the header data that will be added based on the value of `hdr_add_fields`. 

Furthermore the `void *hdr_data` value could possibly be a struct in itself, as an array of unions or something. Clearly each `ULL_ADV_PDU_HDR_FIELD_` bit defines a structured value (e.g. `ULL_ADV_PDU_HDR_FIELD_CTE_INFO` which is just a uint8_t value and `ULL_ADV_PDU_HDR_FIELD_SYNC_INFO` which is more). 

Generally the whole way this is being used is fairly hard to read as code and generally assumes that the reader is intensely familiar with all the values and implicit structs, but I suppose that has been done to try to make it as efficient as posisble?

ppryga replied:
>It actually seems to me that the last 3 arguments (hdr_add_fields, hdr_rem_fields and hdr_data) could benefit from being a struct, or at least hdr_add_fields and hdr_data as, if I understand correct, hdr_data, is the header data that will be added based on the value of hdr_add_fields.

Yes, it could be changed into `struct`, that will add some code in calls like this: 
 https://github.com/zephyrproject-rtos/zephyr/blob/b8aa5b2ecb0410a64044f8119ea220d9d85b4096/subsys/bluetooth/controller/ll_sw/ull_df.c#L313-L315

But that is something we could live with.

> Furthermore the void *hdr_data value could possibly be a struct in itself, as an array of unions or something. Clearly each ULL_ADV_PDU_HDR_FIELD_ bit defines a structured value (e.g. ULL_ADV_PDU_HDR_FIELD_CTE_INFO which is just a uint8_t value and ULL_ADV_PDU_HDR_FIELD_SYNC_INFO which is more).

This is something I'm supporting and I should have thought about it when adding the change to the PR. 
I'm not sure if something as compact as `union` of `structs` is possible here. Pay attention that the goal of the functionality is to allow to provide content of multiple fields at once. 
But switch from compact array into `struct` would increase readability of the code.
There will be small increase of stack usage if we switch to structure holding data as members.

@cvinayak I'd like to hear your opinion about it?

cvinayak replied:
The bit fields and the value array was inspired from how Bluetooth PDU itself is formatted using the header `flag` and then the field values themself thereafter. The interface should permit multiple fields of variable lengths to be populated in a single run of the traversal of the PDU extended header format. 

Passing constant values as parameters reduces indirect memory access which is otherwise required both in the caller and callee when using `structs`.

`union` would restrict passing multiple fields, and `struct` would need worst case allocations, hence the current design of encoding the fields inspired by the extended common header format, with additional optimization to be able to pass multiple fields by reference.

This is one place multiple fields (AdvA, TgtA, ADI, AuxPtr, and TxPower flags) should be passed: https://github.com/zephyrproject-rtos/zephyr/blob/53cd7f7152da476ccfb97ce6171b43c4b195ef99/subsys/bluetooth/controller/ll_sw/ull_adv.c#L532
The code above it has not yet been ported to use `ull_adv_aux_hdr_set_clear`, which is similar to `ull_adv_sync_pdu_set_clear`

ppryga replied:
I'm fine with `hdr_add_fields` and  `hdr_rem_fields` parameters. But in my humble opinion:
```
struct ull_adv_ext_header_data {
    struct pdu_cte_info *cte_info;
    struct {
        uint8_t len;
        uint8_t *data;
    } adv_data ;
    struct {
        uint8_t len;
        uint8_t *ret_data_ptr;
    } acad;
};
```

This would make understanding of the code far easier and lower risk of making a mistake.
The cost is a few bytes on stack + few more instructions due to indirect access. 
@cvinayak @Thalley any further opinions about it?

ppryga replied:
Ok then. Lest make it a policy, at least for controller code,: **"Favour low RAM and ROM usage sometimes sacifying readability."**

@Thalley in such case the code stays in current form.
I'll do rebase then it will be ready for merge.

Thalley replied:
Agreed

Thalley reviewed (approved):
Your changes look sane. I left a comment (which may or may not be addressed here) regarding `ull_adv_sync_pdu_set_clear` as a whole.

Thalley commented:
@ppryga If you add additional code (like you have in this instance), and want to make sure that reviewers re-review it before this PR is merged, you could/should remove their reviews.

carlescufi commented:
@ppryga can you please rebase?

ppryga commented:
@carlescufi The PR is ready for merge
